src/main/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchema.java

```diff
@@ -130,38 +130,44 @@ private static Schema combineArrayValueSchema(
     }
 
     if (firstSchema.type() != secondSchema.type()) {
-      LOGGER.debug(
-          "Can't combine non-matching schema types: {} and {}",
-          firstSchema.type(),
-          secondSchema.type());
+      logIncompatibleSchemas(firstSchema, secondSchema);
       return INCOMPATIBLE_SCHEMA_TYPE;
     }
 
-    if (firstSchema.type() != Schema.Type.STRUCT || secondSchema.type() != Schema.Type.STRUCT) {
-      LOGGER.debug("Can't combine non-equal schema that are not both structs");
-      return INCOMPATIBLE_SCHEMA_TYPE;
-    }
-
-    SchemaBuilder builder = SchemaBuilder.struct().name(firstSchema.name()).optional();
-
-    // _id field first
-    Field id1 = firstSchema.field(ID_FIELD);
-    Field id2 = secondSchema.field(ID_FIELD);
-    if (id1 != null || id2 != null) {
-      builder.field(ID_FIELD, combineFieldSchema(id1, id2));
+    switch (firstSchema.type()) {
+      case ARRAY:
+        SchemaBuilder arrayBuilder =
+            SchemaBuilder.array(
+                    combinedSchema(firstSchema.valueSchema(), secondSchema.valueSchema()))
+                .name(firstSchema.name())
+                .optional();
+        return arrayBuilder.build();
+      case STRUCT:
+        SchemaBuilder structBuilder = SchemaBuilder.struct().name(firstSchema.name()).optional();
+
+        // _id field first
+        Field id1 = firstSchema.field(ID_FIELD);
+        Field id2 = secondSchema.field(ID_FIELD);
+        if (id1 != null || id2 != null) {
+          structBuilder.field(ID_FIELD, combineFieldSchema(id1, id2));
+        }
+        // Combine other fields in name order
+        Stream.concat(
+                firstSchema.fields().stream().map(Field::name),
+                secondSchema.fields().stream().map(Field::name))
+            .filter(name -> !name.equals(ID_FIELD))
+            .distinct()
+            .sorted()
+            .forEach(
+                name ->
+                    structBuilder.field(
+                        name,
+                        combineFieldSchema(firstSchema.field(name), secondSchema.field(name))));
+        return structBuilder.build();
+      default:
+        logIncompatibleSchemas(firstSchema, secondSchema);
+        return INCOMPATIBLE_SCHEMA_TYPE;
```

Comment:
This branch isn't covered by tests.  Can you add one?

Reply:
We shouldn't hit this code - as the schema types have to be equal see line 132

The default is only there for completeness - it could be converted to an assertion and fail but I opted to log and return a valid type.

Reply:
Could we just remove the conditional on line 132 and then this code will cover the condition?  If so that seems clearer and simpler.

Reply:
We have to check both types are the same before switching on the type.  

If `firstSchema` was a Struct and `secondSchema` was a int - then it would fail.

Reply:
Ah ok.  Can you just add a comment like `// Effectively unreachable`

Reply:
Will do